dtm0/dtx.c

```diff
@@ -344,6 +344,13 @@ static void dtx_persistent_ast_cb(struct m0_sm_group *grp,
 
 	M0_ENTRY("dtx=%p, fop=%p", dtx, fop);
 
+	if (M0_IS0(dtx)) {
```

Comment:
I am not sure this is a right solution.
If `dtx` has been cleared, how do you know it is not freed?
How do you know this is not a POISONED pointer?

How can be avoid access this when it is already finalized?
Or how can we disable this callback when it is finalized?

Reply:
> I am not sure this is a right solution.

I agree. We have scheduled rework of dtx and log (https://jts.seagate.com/browse/EOS-21935 and https://jts.seagate.com/browse/EOS-22451). It will be properly addressed.

>  If dtx has been cleared, how do you know it is not freed?  How do you know this is not a POISONED pointer?

The thing is, we do not free() DTXes because we do not free() DTM0 log entries. And we do not free() DTM0 log entries (outside of finalization procedures) because we have not implemented yet the log pruner. So that, it cannot be poisoned, it can be only M0_SET0-ed by dtx_done call.

> How can be avoid access this when it is already finalized? Or how can we disable this callback when it is finalized?

Good point. Yes, we will need this kind of protection once we get log pruner service.  The log has a mutex that protects the list of entries. In a PMA (an object that owns an AST for PERSISTENT message handling within DTX sm group, see m0_dtm0_pmsg_ast) we can store the dtx id (instead of a pointer to dtx); in the ast callback we can lock the log, find the dtx there by id, update it and then release the lock. So that, we indeed have to address this but I guess it will be addressed separately.